### ISSUE TITLE
Tree doesn't expand on loading when expanded is true

### ISSUE THREAD

**Issue by ychen00** [500tech/angular2-tree-component#38]
Any idea why the tree doesn't expand on page loading? Below is my configuration:

In the view inside the Tree element:
 [nodes]="tree" [focused]="true" [options] = "treeOptions"

In the component:
treeOptions = {
        isExpanded: 'expanded',
        treeNodeTemplate: MyTreeNodeTemplate,
        loadingComponent: MyTreeLoadingTemplate,
        context: this
    }

Json returned from the server contains "expanded":true

[{"id":"8365a4e2-698e-11e6-b73e-0242ac12000e","name":"Tom Swift","expanded":true,"children":[{"id":"836da3e3-698e-11e6-b73e-0242ac12000e","name":"[G] User","expanded":true,"children":[{"id":"838d876a-698e-11e6-b73e-0242ac12000e","name":"[S] Launch_Default","expanded":true,"children":[{"id":"83ae0fec-698e-11e6-b73e-0242ac12000e","name":"[P] Launch.refcodes.Put","expanded":true,"children":[]},{"id":"83adf800-698e-11e6-b73e-0242ac12000e","name":"[P] Launch.refcodes.Post","expanded":true,"children":[]},{"id":"83addc0c-698e-11e6-b73e-0242ac12000e","name":"[P] Launch.refcodes.Get","expanded":true,"children":[]},{"id":"83ae25ec-698e-11e6-b73e-0242ac12000e","name":"[P] Launch.refcodes.Delete","expanded":true,"children":[]}]}]},{"id":"836d72ed-698e-11e6-b73e-0242ac12000e","name":"[G] Admin","expanded":true,"children":[{"id":"838c74c5-698e-11e6-b73e-0242ac12000e","name":"[S] IDP_Default","expanded":true,"children":[]},{"id":"838d708a-698e-11e6-b73e-0242ac12000e","name":"[S] Gateway_Default","expanded":true,"children":[]},{"id":"838d876a-698e-11e6-b73e-0242ac12000e","name":"[S] Launch_Default","expanded":true,"children":[{"id":"83ae0fec-698e-11e6-b73e-0242ac12000e","name":"[P] Launch.refcodes.Put","expanded":true,"children":[]},{"id":"83adf800-698e-11e6-b73e-0242ac12000e","name":"[P] Launch.refcodes.Post","expanded":true,"children":[]},{"id":"83addc0c-698e-11e6-b73e-0242ac12000e","name":"[P] Launch.refcodes.Get","expanded":true,"children":[]},{"id":"83ae25ec-698e-11e6-b73e-0242ac12000e","name":"[P] Launch.refcodes.Delete","expanded":true,"children":[]}]},{"id":"838d5849-698e-11e6-b73e-0242ac12000e","name":"[S] OAuth2_Default","expanded":true,"children":[{"id":"ca4e7d44-2a47-4136-a796-04b660299f21","name":"[P] OAuth2.ui.Get","expanded":true,"children":[]}]}]}]}]

**Comment by adamkleingit** [comment]
`isExpandedField: 'expanded'`
instead of:
`isExpanded: 'expanded'`

-- 
- Adam Klein <http://il.linkedin.com/in/adamklein500>
- CTO @ 500Tech.com <http://500tech.com/>
- AngularJS, ReactJS, NodeJS
- Development, consulting, training and public speaking
- (+972) 52 747 5633

**Comment by ychen00** [comment]
I tried isExpandedField: 'expanded' and still it doesn't expand

**Comment by adamkleingit** [comment]
Oh right it's a bug, when loading nodes asynchronously they don't expand
automatically.
Need to fix this.

As a bypass, you can expand the nodes yourself by finding the nodes on the
tree model using getNodeByPath or getNodeById, and calling the expand
method.

-- 
- Adam Klein <http://il.linkedin.com/in/adamklein500>
- CTO @ 500Tech.com <http://500tech.com/>
- AngularJS, ReactJS, NodeJS
- Development, consulting, training and public speaking
- (+972) 52 747 5633

**Comment by tscislo** [comment]
Finally I solved this by adding this method to tree.model.ts
Only below expands parent nodes so the child we try to activate is really visible.
Would you like a PR with that?

`setActiveAndExpandParents(nodeId:string) {
    let active:TreeNode = this.getNodeById(nodeId);
    this.setActiveNode(active, true);

    let parent = active.realParent;
    do {
        this.setExpandedNode(parent, true);
        parent = parent.realParent;
    } while (parent);
}`

**Comment by adamkleingit** [comment]
It sounds a bit too specific. For now just do this from outside the tree.

I planned on adding this to tree-node.model:
```
ensureVisible() {
  if (!this.realParent) return;

  this.realParent.expand();
  this.realParent.ensureVisible();
}
```

So later you could do something like:
```
const node = getNodeById(...)
node.setIsActive(true)
node.ensureVisible()
```

-- 
- Adam Klein <http://il.linkedin.com/in/adamklein500>
- CTO @ 500Tech.com <http://500tech.com/>
- AngularJS, ReactJS, NodeJS
- Development, consulting, training and public speaking
- (+972) 52 747 5633

**Comment by tscislo** [comment]
Your idea is fine. However very often when you use routing you'd like to activate a node and make it visible at the same time. So grouping this in one method would be nice:

``` 
setActiveAndVisible(node) {
const node = getNodeById(node) 
node.setIsActive(true) 
node.ensureVisible() 
}
```

**Comment by adamkleingit** [comment]
I think I can combine setIsActive and ensureVisible to a single function.
Maybe even make ensureVisible a flag that's true by default, because almost
always when you want a node to be active you want it to be visible.

BTW, I forgot we also have to call `scrollIntoView`, to make sure it's
visible.

I will try to get to this in the weekend

-- 
- Adam Klein <http://il.linkedin.com/in/adamklein500>
- CTO @ 500Tech.com <http://500tech.com/>
- AngularJS, ReactJS, NodeJS
- Development, consulting, training and public speaking
- (+972) 52 747 5633

**Comment by adamkleingit** [comment]
I solved it this way:

1) returning `this` after each action method to allow chaining:

node.setIsActive()
  .ensureVisible()
  .scrollIntoView()

2) added setActiveAndVisible(node) which does exactly that

Will update the version soon....

**Comment by tscislo** [comment]
That sounds good! Thx! Waiting for your update.

**Comment by adamkleingit** [comment]
Fixed in 1.2.2
